### ISSUE TITLE
[CT-1852] [Feature] Spike research on dimension `type`

### ISSUE THREAD

**Issue by callum-mcdata** [dbt-labs/dbt-core#6662]
Part of #6626
-----

### What this issue entails?

After some conversations with potential integration partners (and the timeless https://github.com/dbt-labs/dbt-core/discussions/4621), I'm wondering if we should convert the `data_type` property of dimensions into something else. The name is still up in the air but for now let's call it `type`.  

This `type` property could be used to represent that the dimension represents a common semantic type of data. Not it's underlying format (string, int, etc) but its semantic meaning (email, phone number, etc). 

And we can expand that. Maybe `type` has its own level of metadata. For example:
```
type:
   - name: email
     description: "An email address"
     .... (some more semantic info)
```
This could allow users to create their own types and layer them into the project wherever necessary. 

Am I convinced that this is a path we should go down? Absolutely not. But multiple partners have now spoken about the value of understanding that a particular field's semantic meaning beyond its data type so it is something that we should consider.

**Comment by callum-mcdata** [comment]
After a discussion with @MichelleArk , we realized that this might be applicable beyond just `entities`. In our conversation, I noted:

>I wonder if this is an evolution of tags. In the past I’ve used tags to denote things that have to have custom policies applied but that always felt … workaround-y

Should this `type` (or maybe we call it `attribute`) be acceptable across all columns/dimensions? 

### Conclusion
After giving this some thought and drawing up a few mockups, I think we should **refrain from expanding `attributes/types` beyond `entities`.**

**Reasons:**
- The goal for this piece of work is the ability to add more semantic information to a particular dimension. While having this information in the model **could** be value add, it would represent more information that has to be duplicated from model --> model.
  - By this I mean that if these `attributes` were added to any model upstream of the one that the entity is built upon, this information would not be inherited through the DAG unless manually specified. 
- Using `tags` as a mechanism for applying transformation in a model is a **very** advanced use of dbt. This stream of work is focused just on the semantic layer where we want to expand the surface area of dbt to consumers.
- There's nothing stopping us from implementing this in the future for models if people want it!

@MichelleArk does that logic track to you?

**Comment by jtcohen6** [comment]
@MichelleArk @callum-mcdata `tags` or `policy_tags` ([docs](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#policy-tags))?

---

This reminds me of a conversation I've had with Core engineers before, about whether generic tests are really a way of defining custom data types—e.g. `accepted_targets` is (sorta) an enum—just that they're enforced, today, after a model builds.

It also occurred to me that the ongoing work around constraints in support of multi-project is also relevant here. On certain data platforms—today just Postgres + Databricks, maybe more in the future—it's possible to define custom row-level "checks":
```yml
columns:
  - name: my_email_column
    constraints_check: "my_email_column ~ <some_regex>"  # could wrap this logic in an is_valid_email() udf
```
I don't think we should try to get _too_ fancy here. Just that, if we did add cool flexible semantic types, we'd probably need some underlying logical validation to ensure they mean what they say they mean.

**Comment by callum-mcdata** [comment]
I believe the context of our conversation was around `tags`, which were then probably used in macros to enable certain behavior. I know I've used them in the past for things like PII and hashing certain columns if they were tagged.

I think that is why I came to the conclusion that `attributes/types` should probably be different and much lighter scope. Validation and constraints checking are vitally important but I think the context of what this would be aiming for would only be enriching the semantic information of a dimension so that integration partners could more intelligently use them in their applications.

The example given to me most clearly was from our friends over at Census wanting to know what field the email was for a customer entity without needing to try and parse it from column name or from data checks (or manual input).
